### ISSUE TITLE
Think hook jitter

### ISSUE THREAD

**Issue by thegrb93** [Facepunch/garrysmod-issues#3199]
For some reason, things like ApplyForceCenter are not consistent in the Think hook. I've made a plot of d/dt-CurTime vs SysTime and also a video showing this periodic jitter.

Pretty much all E2 creations are affected by this. If we could get it consistent then things would be much better.

The spikes here seem to be periodic every 0.5 sec.
![JitterPlot](https://dl.dropboxusercontent.com/s/cgvctvh6hk2soop/jitter.png?dl=0)

The physics jitters seem to be every 1sec.
https://www.youtube.com/watch?v=W0OYG6K0jjs

**Comment by thegrb93** [comment]
Here's another simple test. The entity should hover without moving at all.

```
local e=Entity(301):GetPhysicsObject()
hook.Add("Think","T",function()
    e:ApplyForceCenter(Vector(0,0,e:GetMass()*600*engine.TickInterval()))
end)
```

**Comment by thegrb93** [comment]
^ I know that the downward descent can't be prevented due to physics iterations in between Thinks, but the net velocity should remain stable at 0 I would think. I plotted the velocity of the prop at each think.

**Comment by thegrb93** [comment]
Okay, I was able to make sense of that plot.

The ramp indicates that ApplyForceCenter is putting slightly too much force on the prop. The spike indicates a frame where no force at all is applied to the prop.

The drop is exactly gravity*engine.TickInterval()

This is happening every 99 frames.

**Comment by thegrb93** [comment]
I haven't tested other forces yet so it could be just an issue with gravity.

**Comment by danielga** [comment]
Physics simulation is not synchronized to the Think hook.

**Comment by thegrb93** [comment]
Yeah I know, but is there a way to get it somewhat consistent? If so, then I think that should be the default setting.

**Comment by thegrb93** [comment]
I just found out this only happens in multiplayer. Singleplayer it's fine.

**Comment by danielga** [comment]
Use the Physics* hooks, this requires you to use scripted entities though.

**Comment by thegrb93** [comment]
Again, I know that physics is seperate from Think, but the fact it's jittering in multiplayer and not in singleplayer is indicative of an issue.

**Comment by G-Force007** [comment]
https://github.com/Facepunch/garrysmod-issues/issues/2502

**Comment by Kefta** [comment]
This isn't related to CurTime imprecision. The jitter is an interval.

**Comment by ZehM4tt** [comment]
Do not use Think for this, use PhysicsSimulate instead and use its specified delta time.

Also keep in mind theres interpolation involved aswell, since theres no clientside physics object and I'm guessing no clientside logic to move the object you can only snap to the last server value.

**Comment by thegrb93** [comment]
The anomolies were measured in serverside.

**Comment by Doliman100** [comment]
https://github.com/Facepunch/garrysmod-issues/issues/3199#issuecomment-303477974, It probably doesn't depend on single-player/multi-player, but on a `-tickrate` argument. The default tick rate matches the physics tick interval 0.015 s, but some servers use `-tickrate 33` which is 0.03(03) s.

https://github.com/Facepunch/garrysmod-issues/issues/3199#issuecomment-302921430, I added game and physics ticks sync to your snippet. Now the prop stays on place with any tick rate if pass the right `game_ticks_offset` input. Read more about how to calculate it here: https://github.com/wiremod/wire/issues/3575
```lua
local game_ticks_offset = 35
local game_tick_interval = engine.TickInterval()
local physics_tick_interval = 0.014999999664723873

local e = Entity(49):GetPhysicsObject()
hook.Add("Think","T",function()
    local game_ticks = engine.TickCount()

    local n1 = math.ceil((game_ticks - game_ticks_offset) * game_tick_interval / physics_tick_interval)
    local n2 = math.ceil((game_ticks - game_ticks_offset + 1) * game_tick_interval / physics_tick_interval)
    local n = n2 - n1

    local v = 600 * physics_tick_interval * (1 + n) / 2
    e:ApplyForceCenter((Vector(0, 0, v) - e:GetVelocity()) * e:GetMass())
end)
```

**Comment by thegrb93** [comment]
That definitely is the most hacky source engine fix I've seen. Looks probably right.
